File: src/main/java/org/example/SendNotificationService.java
package org.example;

import java.util.ArrayList;

public class SendNotificationService {
    private void logging(String message) {
        System.out.println("(Send Notification Service): " + message);
    }
    void run(DatabaseConnection databaseConnection) {
        while (true) {
            ArrayList<SubscriptionData> subscriptionData = databaseConnection.getSubscriptionData();
            for (SubscriptionData data : subscriptionData) {
                logging("Checking for " + data.getUsername());
                String username = data.getUsername();
                String password = data.getPassword();
                BotAdapter botAdapter = new BotAdapter();

                GetChatDataResult newChatData = botAdapter.getChatData(username, password);
                logging("(Bot Service): finished getting chat data");
                String oldChatData = databaseConnection.readChatData(username);
                logging("finished reading chat data");
                if (newChatData.getStatus() != BotResult.SUCCESS) {
                    System.out.println("Error: " + newChatData.getStatus());
                    continue;
                }
                String newChatDataString = newChatData.getChatData();
                assert oldChatData != null;
                if (!oldChatData.equals(newChatDataString)) {
                    String deviceId = data.getDeviceId();
                    String title = "New Chat";
                    String message = "new chat";
                    try {
                        new PushNotification().pushNotification(title, message, deviceId);
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                } else {
                    System.out.println("No new chat");
                }
            }
        }
    }
}

File: src/main/java/org/example/SendNotificationService.kt
package org.example

class SendNotificationService {
    private fun logging(message: String) {
        println("(Send Notification Service): $message")
    }

    fun run(databaseConnection: DatabaseConnection) {
        println("Send Notification Service is running")
        while (true) {
            val subscriptionData = databaseConnection.getSubscriptionData()
            for ((username, password, deviceId) in subscriptionData) {
                logging("Checking for $username")
                val botAdapter = BotAdapter()

                val newChatData = botAdapter.getChatData(username, password)
                logging("(Bot Service): finished getting chat data")
                val oldChatData = databaseConnection.readChatData(username)
                logging("finished reading chat data")
                if (newChatData.status != BotResult.SUCCESS) {
                    println("Error: " + newChatData.status)
                    continue
                }
                val newChatDataString = newChatData.chatData
                val resultStatus = newChatData.status
                if (resultStatus != BotResult.SUCCESS) {
                    println("Error: $resultStatus")
                } else {
                    if (oldChatData == null) {
                        assert(newChatDataString != null)
                        println("username: $username")
                        println("new chat data: $newChatDataString")
                        databaseConnection.saveChatData(username, newChatDataString!!)
                        continue
                    }
                    println("old chat data: $oldChatData")
                    println("new chat data: $newChatDataString")
                    if (oldChatData != newChatDataString) {
                        val title = "New Chat"
                        val message = "new chat"
                        try {
                            PushNotification().pushNotification(title, message, deviceId)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }

                        // save to a database
                        databaseConnection.saveChatData(username, newChatDataString!!)
                    } else {
                        println("No new chat")
                    }
                }
            }
        }
    }
}
